modules/drivers/mongo/src/metabase/driver/mongo/conversion.clj

```diff
@@ -0,0 +1,139 @@
+(ns metabase.driver.mongo.conversion
+  "This namespace contains utilities for conversion between mongo specific types and clojure.
+
+   It is copy of monger's conversion namespace, that was adjusted for the needs of Document. Extensions that were
+   previously implemented in our mongo driver were also moved into this namespace.
+
+   [[to-document]] and [[from-document]] are meant to be used for transformation of clojure data into mongo aggregation
+   pipelines and results back into clojure structures.
+
+   TODO: Logic is copied from monger's conversions. It seems that lot of implementations are redundant. I'm not sure
+         yet. We should consider further simplifying the namespace.
+   TODO: Consider use of bson's encoders/decoders/codecs instead this code.
+   TODO: Or consider adding types from org.bson package -- eg. BsonInt32. If we'd decide to go this way, we could
+         transform ejson completely to clojure structures. That however requires deciding how to represent
+         eg. ObjectIds (it could be eg. {$oid \"...\"} which would copy EJSON v2 way). [EJSON v2 doc](https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/).
+   TODO: Names of protocol functions and protocols are bit misleading as were in monger.
+
+   TODOs should be addressed during follow-up of monger removal."
+  (:require
+   [flatland.ordered.map :as ordered-map]
+   [java-time.api :as t]
+   [metabase.query-processor.timezone :as qp.timezone]))
+
+(set! *warn-on-reflection* true)
+
+;;;; Protocols defined originally in monger, adjusted for `Document` follow.
+
+(defprotocol ConvertFromDocument
+  (from-document [input opts] "Converts given DBObject instance to a piece of Clojure data"))
+
+(extend-protocol ConvertFromDocument
+  nil
+  (from-document [input _opts] input)
+
+  Object
+  (from-document [input _opts] input)
+
+  org.bson.types.Decimal128
+  (from-document [^org.bson.types.Decimal128 input _opts]
+    (.bigDecimalValue input))
+
+  java.util.List
+  (from-document [^java.util.List input opts]
+    (mapv #(from-document % opts) input))
+
+  java.util.Date
+  (from-document [t _]
+                 (t/instant t))
+
+  org.bson.Document
+  (from-document [input {:keys [keywordize] :or {keywordize true} :as opts}]
+    (reduce (if keywordize
+              (fn [m ^String k]
+                (assoc m (keyword k) (from-document (.get input k) opts)))
+              (fn [m ^String k]
+                (assoc m k (from-document (.get input k) opts))))
+            (ordered-map/ordered-map)
+            (.keySet input))))
+
+(defprotocol ConvertToDocument
+  (to-document [input] "Converts given piece of Clojure data to org.bson.Document usable by java driver."))
+
+(extend-protocol ConvertToDocument
+  nil
+  (to-document [_input] nil)
+
+  String
+  (to-document [input] input)
+
+  Boolean
+  (to-document [input] input)
+
+  java.util.Date
+  (to-document [input] input)
```

Comment:
These look quite superfluous given that they are `Object`s and they don't change the behavior.